concordium-consensus/src/Concordium/KonsensusV1/Scheduler.hs

```diff
@@ -402,10 +415,75 @@ processBlockRewards ParticipatingBakers{..} TransactionRewardParameters{..} miss
   where
     hasValidatorSuspension = sSupportsValidatorSuspension (sAccountVersionFor (protocolVersion @pv))
 
+-- | Suspend all validators with more missed rounds than the threshold specified
+--  in the chain parameters. Logs the suspension of a validator in a special
+--  transaction outcome.
+processSuspensions ::
+    forall pv m.
+    ( pv ~ MPV m,
+      BlockStateStorage m,
+      IsConsensusV1 pv,
+      PVSupportsValidatorSuspension pv
+    ) =>
+    Timestamp ->
+    UpdatableBlockState m ->
+    m (UpdatableBlockState m)
+processSuspensions timestamp bs0 = do
+    -- get chain paramers
+    ps <- bsoGetChainParameters bs0
+    case _cpValidatorScoreParameters ps of
+        NoParam -> return bs0 -- Nothing to do here.
+        SomeParam (ValidatorScoreParameters{..}) -> do
+            -- get baker pool reward details
+            bprd <- bsoGetBakerPoolRewardDetails bs0
+            -- account indexes that will be suspended
+            let ixs =
+                    [ bakerAccountIndex bid
+                      | (bid, rd) <- Map.toList bprd,
+                        fromCondDef (suspensionPrimed rd) False,
+                        fromCondDef (missedRounds rd) 0 > _vspMaxMissedRounds
+                    ]
+            (upds, bs1) <-
+                foldM
+                    ( \(upds, bs) accIx -> do
+                        res <- bsoUpdateValidator bs timestamp accIx updSuspend
+                        case res of
+                            -- TODO(drsk) how to deal with error?
+                            Left _err -> return (upds, bs)
+                            Right (upds', bs') -> return (upds ++ upds', bs')
```

Comment:
I would consider introducing a new operation `bsoSuspendValidator` that just suspends the validator, so that you don't have to worry about impossible error cases here. I'm also a bit suspicious that the precondition that the account is actually a validator may not be met here, because it could be that the validator has been removed since the last snapshot. `bsoSuspendValidator` could be safe to call on accounts that are not validators, and even return whether the validator was actually suspended. When we add the special outcomes, we can probably limit to those that were actually suspended, leaving out any that had already been manually suspended or removed since the last snapshot.

Reply:
Implemented a `bsoSuspendValidator` block state operation.